### PR TITLE
revert "fix wrong output when using jsonpath"

### LINKED ISSUE #104276: kubectl get -o jsonpath no longer continues parsing an item when a key is encountered that has no value
#### What happened:

`kubectl get -o jsonpath` no longer continues parsing an item when a key is encountered that has no value.  This appears to be a behavior change introduced in Kubernetes version 1.22.

#### What you expected to happen:

`kubectl get -o jsonpath` will continue .

#### How to reproduce it (as minimally and precisely as possible):

Run the following commands against a cluster running CoreDNS pods in the `kube-system` namespace:

```
raw_run_as_user_results=$(kubectl get pods -n kube-system -l k8s-app=kube-dns -o jsonpath='{range .items[*]}{"SED_NEW_LINE_FILTER"}{"podName="}{.metadata.name}{":podNamespace="}{.metadata.namespace}{":podRunAsUser="}{.spec.securityContext.runAsUser}{":podRunAsNonRoot="}{.spec.securityContext.runAsNonRoot}{":"}{range .spec.containers[*]}{"containerName="}{.name}{":containerRunAsUser="}{.securityContext.runAsUser}{":containerRunAsNonRoot="}{.securityContext.runAsNonRoot}{":containerPrivileged="}{.securityContext.privileged}{":"}{end}{end}')
formatted_run_as_user_results=$(sed 's/SED_NEW_LINE_FILTER/\n/g' <<< "${raw_run_as_user_results}")
echo "${formatted_run_as_user_results}"
```

#### Anything else we need to know?:

Nothing.

#### Environment:
- Kubernetes version (use `kubectl version`): v1.22.0
- Cloud provider or hardware configuration: N/A
- OS (e.g: `cat /etc/os-release`): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Network plugin and version (if this is a network-related bug): N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by atiratree**
#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:
This partially reverts commit #98057

the original problem was caused by not using {end} at the end of the range. Please see attached jsonpath documentation on using range.

The reverted commit causes a new bug of not parsing missing keys incorrectly (added a unit test for this use case)

#### Which issue(s) this PR fixes:

Fixes #104276

#### Does this PR introduce a user-facing change?

```release-note
Fix a 1.22 regression with wrong output using jsonpath
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[Other doc]: https://kubernetes.io/docs/reference/kubectl/jsonpath/
```

### CODE REVIEW

**Comment by atiratree** [comment]
@zerodayz @soltysh

**Review by knight42 — APPROVED** [review]
/ok-to-test
/lgtm

Thanks for the fix!

**Comment by knight42** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/104172/pull-kubernetes-integration/1423531656039370752

These errors doesn't seem to be legit.
/retest

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
@atiratree can you open a cherry-pick against 1.22 to fix this problem since that's where the regression was introduced.

**Comment by soltysh** [comment]
ping @caesarxuchao for approval

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go [hunk 1]
```diff
@@ -806,8 +807,8 @@ func TestRunningPodsJSONPathOutput(t *testing.T) {
 	testJSONPath(
 		[]jsonpathTest{
 			{
-				"when range is used in a certain way in script, additional line is printed",
-				`{range .items[?(.status.phase=="Running")]}{.metadata.name}{" is Running\n"}`,
+				"range over pods without selecting the last one",
+				`{range .items[?(.status.phase=="Running")]}{.metadata.name}{" is Running\n"}{end}`,
```
Comment (sttts) [inline staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go:811]: shouldn't we have both cases tested, ie with and without end?

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go [hunk 2]
```diff
@@ -280,6 +280,7 @@ func TestStructInput(t *testing.T) {
 
 	missingKeyTests := []jsonpathTest{
 		{"nonexistent field", "{.hello}", storeData, "", false},
+		{"nonexistent field 2", "before-{.hello}after", storeData, "before-after", false},
```
Comment (sttts) [inline staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go:283]: how is this related? I don't see it in the reverted PR.

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go (hunk 1 above)
Comment (atiratree) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go:811]: Should we test invalid syntaxes here? What should be the behaviour?
- the output is broken and outputs additional line (current behaviour)
- or should we fast fail with invalid syntax error? (would require changes)

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go (hunk 2 above)
Comment (atiratree) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go:283]: this is a behaviour that got broken by the reverted commit. Ie the template got discarded after the `{.hello}` element.

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go:811]: it was kind of graceful before to leave out the `{end}`, just a little wrong. Could be get it working easily too?

Comment (sttts) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go:811]: In any case, I would rather try to avoid breaking behaviour that ran without error before. Fixing it is ok, but not suddenly fail.

Comment (atiratree) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go:811]: I am not sure how easy that would be. 
But do we want to support an invalid syntax? I think it is a bit problematic of supporting this especially when you consider nested ranges. I would prefer an error here , but that would also break some scripts (more than keeping it failing slightly).

Comment (soltysh) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go:811]: I think that the support for missing `{end}` was rather accidental than intentional. I'd prefer we stick with having code simple and complying to jsonpath standard, than over-complicated to support something not entirely correct.

**Review by soltysh — APPROVED** [review]
/lgtm

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go:811]: I would prefer keeping graceful behaviour and not fix it and break users. But I am not blocking this. Both paths are acceptable.

**Comment by sttts** [comment]
/approve

with https://github.com/kubernetes/kubernetes/pull/104172#discussion_r699239375 in mind.

**Comment by atiratree** [comment]
the failures don't seem connected to this PR
/retest

**Comment by rtheis** [comment]
If not too much trouble, can we link this PR as the fix for https://github.com/kubernetes/kubernetes/issues/104276? Thank you.

**Comment by soltysh** [comment]
> If not too much trouble, can we link this PR as the fix for #104276? Thank you.

done
